Challenge-4/Bot_2/src/agent.spec.ts

```diff
@@ -0,0 +1,459 @@
+      inputs: [],
+      outputs: [utilization],
+    });
+    mockProvider.addCallTo(usdcTokenAddress, 0, Iface, "getSupplyRate", {
+      inputs: [utilization],
+      outputs: [986453221],
+    });
+    mockProvider.addCallTo(usdcTokenAddress, 0, Iface, "getBorrowRate", {
+      inputs: [utilization],
+      outputs: [1064532211],
+    });
+  };
+
+  it("returns findings if supplier liquidates when supply kink is higher that upper limit", async () => {
+    setupMockProvider(
+      ethers.BigNumber.from("860000000000000000"),
+      ethers.BigNumber.from("5000000000000000000"),
+      ethers.BigNumber.from("860000000000000000"),
+    );
+
+    mockProvider.setNetwork(1);
+    await initialize();
+    txEvent.addTraces({
+      function: provideInterface.getFunction("supply"),
+      to: createAddress("0xc3d688B66703497DAA19211EEdff47f25384cdc3"),
+      from: createAddress("0x123"),
+      arguments: ["0xc3d688B66703497DAA19211EEdff47f25384cdc3", 20],
+    });
+
+    const findings = await handleTransaction(txEvent);
+
+    expect(findings).toStrictEqual([
+      Finding.fromObject({
+        name: `Lender's incentivized to supply`,
+        description: `The Supply APR is 3.11, which is favourable for lenders, as the utlization rate is 86`,
+        alertId: "SUPPLY-1",
+        severity: FindingSeverity.Info,
+        type: FindingType.Info,
+        protocol: "Compound",
+        metadata: {
+          SupplyRate: "3.11",
+          Utilization: "86",
+        },
+      }),
+    ]);
+  });
+
+  // Test for borrower incentive and if it is below the lower limit
+
+  it("returns findings if borrower borrows when borrow kink is less that lower limit", async () => {
+    setupMockProvider(
+      ethers.BigNumber.from("900000000000000000"),
+      ethers.BigNumber.from("9600000000000000000"),
+      ethers.BigNumber.from("300000000000000000"),
+    );
+
+    mockProvider.setNetwork(1);
+    await initialize();
+    txEvent.addTraces({
+      function: provideInterface.getFunction("withdraw"),
+      to: createAddress("0xc3d688B66703497DAA19211EEdff47f25384cdc3"),
+      from: createAddress("0x123"),
+      arguments: ["0xc3d688B66703497DAA19211EEdff47f25384cdc3", 10],
+    });
+
+    const findings = await handleTransaction(txEvent);
+
+    expect(findings).toStrictEqual([
+      Finding.fromObject({
+        name: `Borrower's incentivized to borrow`,
+        description: `The Borrow APR is 3.36, which is favourable for borrowers, as the optimal utilization rate is 30`,
+        alertId: "BORROW-1",
+        severity: FindingSeverity.Info,
+        type: FindingType.Info,
+        protocol: "Compound",
+        metadata: {
+          BorrowRate: "3.36",
+          Utilization: "30",
+        },
+      }),
+    ]);
+  });
+
+  it("returns a finding among multiple transactions that are not supply, if the utlization is above the upper limit of supply kink ", async () => {
```

Comment:
Technically not multiple transactions but multiple function calls. And needs to be rephrased since one is a `supply` and the other is not.